tests/index.test.ts

```diff
@@ -55,4 +56,200 @@ describe('index', () => {
+        this: T,
+        json: any,
+      ): InstanceType<T> {
+        if (
+          typeof json !== 'object' ||
+          json.type !== this.name ||
+          typeof json.data !== 'object' ||
+          typeof json.data.message !== 'string' ||
+          typeof json.data.num !== 'number' ||
+          ('stack' in json.data && typeof json.data.stack !== 'string')
+        ) {
+          throw new TypeError(`Cannot decode JSON to ${this.name}`);
+        }
+        const e = new this(json.data.num, json.data.message);
+        e.stack = json.data.stack;
+        return e;
+      }
+      public num: number;
+      public constructor(num: number, message: string) {
+        super(message);
+        this.num = num;
+      }
+      public toJSON() {
+        const obj = super.toJSON();
+        obj.data.num = this.num;
+        return obj;
+      }
+    }
+    class UnknownError<T> extends AbstractError<T> {}
+    // AbstractError classes, these should be part of our application stack
+    const customErrors = {
+      AbstractError,
+      SpecificError,
+      UnknownError,
+    };
+    // Standard JS errors, these do not have fromJSON routines
+    const standardErrors = {
+      Error,
+      TypeError,
+      SyntaxError,
+      ReferenceError,
+      EvalError,
+      RangeError,
+      URIError,
+    };
+    function replacer(key: string, value: any): any {
+      if (value instanceof Error) {
+        return {
+          type: value.constructor.name,
+          data: {
+            message: value.message,
+            stack: value.stack,
+          },
+        };
+      } else {
+        return value;
+      }
+    }
+    // Assume that the purpose of the reviver is to deserialise JSON string
+    // back into exceptions
+    // The reviver has 3 choices when encountering an unknown value
+    // 1. throw an "parse" exception
+    // 2. return the value as it is
+    // 3. return as special "unknown" exception that contains the unknown value as data
+    // Choice 1. results in strict deserialisation procedure (no forwards-compatibility)
+    // Choice 2. results in ambiguous parsed result
+    // Choice 3. is the best option as it ensures a typed-result and debuggability of ambiguous data
+    function reviver(key: string, value: any): any {
+      if (
+        typeof value === 'object' &&
+        typeof value.type === 'string' &&
+        typeof value.data === 'object'
+      ) {
+        try {
+          let eClass = customErrors[value.type];
+          if (eClass != null) return eClass.fromJSON(value);
+          eClass = standardErrors[value.type];
+          if (eClass != null) {
+            if (
+              typeof value.data.message !== 'string' ||
+              ('stack' in value.data && typeof value.data.stack !== 'string')
+            ) {
+              throw new TypeError(`Cannot decode JSON to ${value.type}`);
+            }
+            const e = new eClass(value.data.message);
```

Comment:
Demonstrate the parsing of `cause` and setting of `cause` for node v16.

Also `AggregateError` has it's own array of errors too.

Reply:
Done this for `AggregateError`.